### PR TITLE
Ability to set a custom namespace for the redis keys.

### PR DESCRIPTION
**Pull request by rberrelleza**
This is based on the work we tried before on https://github.com/nvie/rq/pull/459 (It was easier to create a new PR than rebase and all that). I took all the comments you had, plus Jay's original idea, and re-implemented it.

Some testing is still needed, but wanted to get your thoughts on this approach before going further down this path.

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from rq.utils import build_key
+from tests import RQTestCase
+
+
+class TestHelpers(RQTestCase):
+
+    def test_build_key(self):
+        """Ensure the redis key is correctly build when there's a namespace present"""
```
Comment (tomviner) [inline tests/test_utils.py]: I think you mean *is correctly built*

File: rq/queue.py [hunk 2]
```diff
@@ -32,33 +32,35 @@ class Queue(object):
...
         the internal Redis keys.  Can be used to reverse-lookup Queues by their
         Redis keys.
         """
-        prefix = cls.redis_queue_namespace_prefix
+        prefix = build_key(cls.redis_queue_namespace_prefix, namespace)
+
         if not queue_key.startswith(prefix):
             raise ValueError('Not a valid RQ queue key: {0}'.format(queue_key))
         name = queue_key[len(prefix):]
         return cls(name, connection=connection)
```
Comment (olruzhytskyi) [inline rq/queue.py]: There should be a namespace=namespace positional argument to create an instance with namespace.

**Comment by torfsen** [comment]
This feature would come in very handy for us -- is there anything I can do to help it along?

**Comment by proofit404** [comment]
Why anyone want to customize such internal stuff of rq?

**Comment by TkTech** [comment]
👍 this would be of great help to the ckan/ckan project.

**Comment by proofit404** [comment]
I'm sorry I still don't get it. Can you elaborate more on the term "multiple simultaneous job queues". What problems does this pattern solve? In which way it do that?

**Comment by rberrelleza** [comment]
@proofit404 in our case, we have different services that use RQ in the same redis instance.  Currently, rq assumes that the rq:{id} or rq:{queue_name} are exclusive, which doesn't work well in the described scenario, mainly due to name collisions.

**Comment by proofit404** [comment]
@rberrelleza Why you use one database for all you services? It's simple to use database 0 for service A and database 1 for service B. Is this appropriate solution to you?

**Comment by torfsen** [comment]
@proofit404 Many small/free plans for hosted Redis only include a single database.

**Comment by proofit404** [comment]
@torfsen Ok. Different databases does not work in this case.

@rberrelleza As rq documentation says you must provide unique job ids (if you use don't use uuid module) and unique queue names for all your running services. The only possible name collision not under your control is failing queues. You can override Worker.move_to_failed_queue and FailedQueue.__init__ if you want to separate failures from different services.

At current state rq intended to be "one cluster per one redis instance". If you want to change this paradigm into "multiple clusters per one redis instance" then add explicit API part for this and leave calculated prefixes as internal part.

**Pull request closed without merging**